File: splay_tree/SplayTree.java
/**
 * Implements a top-down splay tree.
 * Available at http://www.link.cs.cmu.edu/splay/
 * Author: Danny Sleator <sleator@cs.cmu.edu>
 * This code is in the public domain.
 */

class BinaryNode
{
    BinaryNode(Comparable theKey) {
        key = theKey;
        left = right = null;
    }

    Comparable key;          // The data in the node
    BinaryNode left;         // Left child
    BinaryNode right;        // Right child
}

public class SplayTree
{
    private BinaryNode root;

    public SplayTree() {
        root = null;
    }

    /**
     * Insert into the tree.
     * @param x the item to insert.
     * @throws DuplicateItemException if x is already present.
     */
    public void insert(Comparable key) {
	BinaryNode n;
	int c;
	if (root == null) {
	    root = new BinaryNode(key);
	    return;
	}
	splay(key);
	if ((c = key.compareTo(root.key)) == 0) {
	    //	    throw new DuplicateItemException(x.toString());	    
	    return;
	}
	n = new BinaryNode(key);
	if (c < 0) {
	    n.left = root.left;
	    n.right = root;
	    root.left = null;
	} else {
	    n.right = root.right;
	    n.left = root;
	    root.right = null;
	}
	root = n;
    }

    /**
     * Remove from the tree.
     * @param x the item to remove.
     * @throws ItemNotFoundException if x is not found.
     */
    public void remove(Comparable key) {
	BinaryNode x;
	splay(key);
	if (key.compareTo(root.key) != 0) {
	    //            throw new ItemNotFoundException(x.toString());
	    return;
	}
	// Now delete the root
	if (root.left == null) {
	    root = root.right;
	} else {
	    x = root.right;
	    root = root.left;
	    splay(key);
	    root.right = x;
	}
    }

    /**
     * Find the smallest item in the tree.
     */
    public Comparable findMin() {
        BinaryNode x = root;
        if(root == null) return null;
        while(x.left != null) x = x.left;
        splay(x.key);
        return x.key;
    }

    /**
     * Find the largest item in the tree.
     */
    public Comparable findMax() {
        BinaryNode x = root;
        if(root == null) return null;
        while(x.right != null) x = x.right;
        splay(x.key);
        return x.key;
    }

    /**
     * Find an item in the tree.
     */
    public Comparable find(Comparable key) {
	if (root == null) return null;
	splay(key);
        if(root.key.compareTo(key) != 0) return null;
        return root.key;
    }

    /**
     * Test if the tree is logically empty.
     * @return true if empty, false otherwise.
     */
    public boolean isEmpty() {
        return root == null;
    }

    /** this method just illustrates the top-down method of
     * implementing the move-to-root operation 
     */
    private void moveToRoot(Comparable key) {
	BinaryNode l, r, t, y;
	l = r = header;
	t = root;
	header.left = header.right = null;
	for (;;) {
	    if (key.compareTo(t.key) < 0) {
		if (t.left == null) break;
		r.left = t;                                 /* link right */
		r = t;
		t = t.left;
	    } else if (key.compareTo(t.key) > 0) {
		if (t.right == null) break;
		l.right = t;                                /* link left */
		l = t;
		t = t.right;
	    } else {
		break;
	    }
	}
	l.right = t.left;                                   /* assemble */
	r.left = t.right;
	t.left = header.right;
	t.right = header.left;
	root = t;
    }

    private static BinaryNode header = new BinaryNode(null); // For splay
    
    /**
     * Internal method to perform a top-down splay.
     * 
     *   splay(key) does the splay operation on the given key.
     *   If key is in the tree, then the BinaryNode containing
     *   that key becomes the root.  If key is not in the tree,
     *   then after the splay, key.root is either the greatest key
     *   < key in the tree, or the lest key > key in the tree.
     *
     *   This means, among other things, that if you splay with
     *   a key that's larger than any in the tree, the rightmost
     *   node of the tree becomes the root.  This property is used
     *   in the delete() method.
     */

    private void splay(Comparable key) {
	BinaryNode l, r, t, y;
	l = r = header;
	t = root;
	header.left = header.right = null;
	for (;;) {
	    if (key.compareTo(t.key) < 0) {
		if (t.left == null) break;
		if (key.compareTo(t.left.key) < 0) {
		    y = t.left;                            /* rotate right */
		    t.left = y.right;
		    y.right = t;
		    t = y;
		    if (t.left == null) break;
		}
		r.left = t;                                 /* link right */
		r = t;
		t = t.left;
	    } else if (key.compareTo(t.key) > 0) {
		if (t.right == null) break;
		if (key.compareTo(t.right.key) > 0) {
		    y = t.right;                            /* rotate left */
		    t.right = y.left;
		    y.left = t;
		    t = y;
		    if (t.right == null) break;
		}
		l.right = t;                                /* link left */
		l = t;
		t = t.right;
	    } else {
		break;
	    }
	}
	l.right = t.left;                                   /* assemble */
	r.left = t.right;
	t.left = header.right;
	t.right = header.left;
	root = t;
    }

    // test code stolen from Weiss
    public static void main(String [ ] args)
    {
        SplayTree t = new SplayTree();
        final int NUMS = 40000;
        final int GAP  =   307;

        System.out.println("Checking... (no bad output means success)");

        for(int i = GAP; i != 0; i = (i + GAP) % NUMS)
            t.insert(new Integer(i));
        System.out.println("Inserts complete");

        for(int i = 1; i < NUMS; i+= 2)
            t.remove(new Integer(i));
        System.out.println("Removes complete");

        if(((Integer)(t.findMin())).intValue() != 2 ||
            ((Integer)(t.findMax())).intValue() != NUMS - 2)
            System.out.println("FindMin or FindMax error!");

        for(int i = 2; i < NUMS; i+=2)
            if(((Integer)t.find(new Integer(i))).intValue() != i)
                System.out.println("Error: find fails for " + i);

        for(int i = 1; i < NUMS; i+=2)
            if(t.find(new Integer(i))  != null)
                System.out.println("Error: Found deleted item " + i);
    }

}

File: splay_tree/go/splay_tree.go
package splay_tree

type SplayTree struct {
	root *node
}

type Data struct {
	Key   int
	Value interface{}
}

type node struct {
	data  Data
	left  *node
	right *node
}

type dataCallback func(Data)

func NewSplayTree() *SplayTree {
	return &SplayTree{nil}
}

func (t *SplayTree) Get(key int) (value interface{}, ok bool) {
	t.root = splay(t.root, key)
	if t.root == nil || key != t.root.data.Key {
		return 0, false
	}
	return t.root.data.Value, true
}

func (t *SplayTree) Set(key int, value interface{}) {
	t.root = insert(t.root, key, value)
}

func (t *SplayTree) Del(key int) {
	t.root = remove(t.root, key)
}

func (t *SplayTree) All() []Data {
	a := []Data{}
	inOrder(t.root, func(d Data) {
		a = append(a, d)
	})
	return a
}

func insert(n *node, key int, value interface{}) *node {
	n = splay(n, key)
	if n != nil && key == n.data.Key {
		n.data.Value = value
		return n
	}
	newNode := &node{Data{key, value}, nil, nil}
	if n == nil {
		return newNode
	}
	if key < n.data.Key {
		// n.data.Key is the smallest key > key in the tree, which guarantees n.left.data.Key < key.
		newNode.left = n.left
		newNode.right = n
		n.left = nil
	} else {
		// Symmetrical case: n.data.Key is the greatest key < key in the tree, which guarantees
		// n.right.data.Key > key.
		newNode.right = n.right
		newNode.left = n
		n.right = nil
	}
	return newNode
}

func remove(n *node, key int) *node {
	n = splay(n, key)
	if n == nil || key != n.data.Key {
		return n
	}
	if n.left == nil {
		return n.right
	}
	if n.right == nil {
		return n.left
	}
	// n has two children, so after deleting n we have two subtrees: n.left and n.right. Steps:
	// 1. Splay the maximum node of the left subtree. This can be done by simply calling
	//    splay(n.left, key), since we know that key > any key in n.left.
	// 2. The root of left subtree now has no right child, since it's the maximum node of the left
	//    subtree. So we can simply graft the right subtree there.
	l := splay(n.left, key)
	l.right = n.right
	return l
}

// splay performs a top-down splay on the given key, and returns the new root of the tree. If key
// is in the tree, the new root is the node with the key. If key is not in the tree, the new root is
// either the node with the greatest key < key in the tree, or the node with the smallest key > key
// in the tree.
func splay(n *node, key int) *node {
	if n == nil {
		return nil
	}

	var leftRoot, leftLast, rightRoot, rightLast *node

	for {
		if key == n.data.Key {
			break
		}
		if key < n.data.Key {
			if n.left != nil && key < n.left.data.Key {
				// Zig-zig.
				n = rotateRight(n)
			}
			if n.left == nil {
				break
			}
			rightRoot, rightLast = linkRight(rightRoot, rightLast, n)
			n = n.left
		} else {
			if n.right != nil && key > n.right.data.Key {
				// Zig-zig.
				n = rotateLeft(n)
			}
			if n.right == nil {
				break
			}
			leftRoot, leftLast = linkLeft(leftRoot, leftLast, n)
			n = n.right
		}
	}

	// Assemble.
	n.right, _ = linkRight(rightRoot, rightLast, n.right)
	n.left, _ = linkLeft(leftRoot, leftLast, n.left)

	return n
}

func linkLeft(root, last, n *node) (newRoot, newLast *node) {
	if root == nil {
		newRoot = n
	} else {
		newRoot = root
		last.right = n
	}
	newLast = n
	return
}

func linkRight(root, last, n *node) (newRoot, newLast *node) {
	if root == nil {
		newRoot = n
	} else {
		newRoot = root
		last.left = n
	}
	newLast = n
	return
}

func rotateLeft(n *node) *node {
	r := n.right
	n.right = r.left
	r.left = n
	return r
}

func rotateRight(n *node) *node {
	l := n.left
	n.left = l.right
	l.right = n
	return l
}

func inOrder(n *node, cb dataCallback) {
	if n == nil {
		return
	}
	inOrder(n.left, cb)
	cb(n.data)
	inOrder(n.right, cb)
}
